src/provider.toit

```diff
@@ -0,0 +1,166 @@
+        // Remove it and replace it with a fresh one.
+        // However, the new watchdog will be in the same state as the old one.
+        old-dog := dogs_[id]
+        dogs_.remove id
+        dog = Watchdog.clone this client old-dog
+      else:
+        dog = Watchdog this client id
+      dogs_[id] = dog
+      return dog
+
+    if index == WatchdogService.START-INDEX:
+      dog := (this.resource client arguments[0]) as Watchdog
+      dog.start (arguments[1] as int)
+      start-system-watchdog-if-necessary_
+      return null
+
+    if index == WatchdogService.FEED-INDEX:
+      dog := (this.resource client arguments[0]) as Watchdog
+      dog.feed
+      return null
+
+    if index == WatchdogService.STOP-INDEX:
+      dog := (this.resource client arguments[0]) as Watchdog
+      dog.stop
+      stop-system-watchdog-if-possible_
+      return null
+
+    unreachable
+
+  start-system-watchdog-if-necessary_:
+    if system-watchdog-task_: return
+
+    mutex_.do:
+      logger_.debug "starting system watchdog"
+      esp32.watchdog-init --ms=GRANULARITY-MS
+      system-watchdog-task_ = task::
+        while true:
+          too-late := false
+          dogs_.do --values: | dog/Watchdog |
+            if dog.is-too-late:
+              logger_.fatal "watchdog too late" --tags={ "id": dog.id }
+              too-late = true
+
+          if too-late:
+            esp32.deep-sleep Duration.ZERO
+          else:
+            // Feed the system watchdog.
+            logger_.debug "feeding system watchdog"
+            mutex_.do: esp32.watchdog-reset
+            sleep --ms=1_000
+
+  stop-system-watchdog-if-possible_:
+    if not system-watchdog-task_: return
+
+    needs-watching := dogs_.any: | dog/Watchdog | not dog.is-stopped
+    if needs-watching: return
+
+    // Shutdown the system watchdog.
+    mutex_.do:
+      logger_.debug "stopping system watchdog"
+      esp32.watchdog-deinit
+
+      system-watchdog-task_.cancel
+      system-watchdog-task_ = null
+
+  remove-dog_ dog/Watchdog:
+    logger_.debug "removing watchdog" --tags={ "id": dog.id }
+    dogs_.remove dog.id
+
+class Watchdog extends ServiceResource:
+  /** The watchdog is started. */
+  static STATE-STARTED ::= 0
+  /** The watchdog is stopped. */
+  static STATE-STOPPED ::= 1
+  /**
+  The watchdog wasn't fed in time.
+
+  Feeding or stopping at this point is too late.
+  */
+  static STATE-TOO-LATE ::= 2
+
+  provider/WatchdogServiceProvider
+  id/string
+
+  state/int := STATE-STOPPED
+  last-feeding-us/int := -1
+  max-sleep-us/int := -1
+
+  constructor .provider client/int .id:
+    super provider client
+
+  constructor.clone .provider client/int other/Watchdog:
+    id = other.id
+    state = other.state
+    last-feeding-us = other.last-feeding-us
+    max-sleep-us = other.max-sleep-us
+
+    super provider client
+
+  start s/int:
+    if not is-too-late:
+      state = STATE-STARTED
+      last-feeding-us = Time.monotonic-us
```

Comment:
Maybe only set this if it wasn't already set? If you create/start successfully, but crash right after that in a loop, it would be nice if the watchdog timer eventually expires and you get a reboot.

Reply:
done.